---
title: "Bulk RNA-Seq analysis"
author: "Elisabeth F. Heuston"
output: html_notebook
---

Analysis of mouse single cell hematopoietic populations - Bulk RNA-Seq differential expression anaysis

This notebook contains the R workflow to analyze bulk RNS-Seq data and pertains to:
* Figure 6A
* Supplemental Figure 7A

Raw data for this publication are available from the GEO Project GSE168260 at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE168260  
Raw data previously published in Heuston et al., 2018, is available at www.usevision.org

## Principal component analysis of hematopoietic RNA-Seq profiles

Transcriptional comparison of 11 populations. Each sample was sequenced in duplicate.  

* LSK  
* CMP  
* GMP  
* CFU-E  
* CFU-Mk  
* EB  
* MK  
* C10 (CMP subpopulation from cluster 10)  
* C11 (CMP subpopulation from cluster 11)  
* C17 (CMP subpopulation from cluster 17)  
* C3 (CMP subpopulation from cluster 3)  

Transcriptome alignments were performed in STAR, and read estimation performed in RSEM. 

### Load libraries

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = '~/Desktop/scRNAanalysis_codetest/RSEM_resultsFiles/')
```

```{r, results=FALSE, message=FALSE, warning=FALSE}
library(tximport)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(edgeR)
library(colorRamps)
library(org.Mm.eg.db)
library(ggplot2)
library(scatterplot3d)
library(mouseSingleCell)
```
### Comprison of all hematopoietic populations
Start by comparing previously isolated bulk RNA and prospectively isolated CMP subpopulations
#### Import read estimation (RSEM) files
The `mouseSingleCell` library contains code, functions, and data to reproduce the analyses presented in the manuscript  
First generate a list of the files to be imported
```{r}
file_list <- list.files(pattern = 'genes.results', ignore.case = FALSE, recursive = FALSE, include.dirs = FALSE)
names(file_list) <- lapply(file_list, function(x) gsub('.genes.results', '', x))
file_list
```
Now import the files using `tximport` to identify the counts columns
```{r, warning=FALSE}
mstxi <- tximport::tximport(files = file_list, type = 'rsem', txIn = FALSE, txOut = FALSE, geneIdCol = 3)
cts <- mstxi$counts
colnames(cts)
```
#### Generate DGE object and QC
Set CMP as reference sample and generate DGE object
```{r}
dgelist_groups <- factor(c('C10', 'C10', 'C11', 'C11', 'C17', 'C17', 'C3s', 'C3s', 'CFUE', 'CFUE', 'CFUMK', 'CFUMK', 'CMP', 'CMP', 'ERY', 'ERY', 'GMP', 'GMP', 'LSK', 'LSK', 'MK', 'MK'))
dgelist_groups <- relevel(dgelist_groups, ref = "CMP")
dgelist <- DGEList(counts = cts, group = dgelist_groups)
apply(dgelist$counts, 2, sum)
```
Keep only genes that have a CPM of 5 in at least 2 samples
```{r}
keep <- rowSums(cpm(dgelist) > 5) >=2
dgelist <- dgelist[keep,]
```
#### ID conversion
Annotate with gene symbols. Note that mouseIDTable is included to permit offline analysis. For up-to-date annotations, use `mouseSingleCell::Mouse2Human_idconversion`, which accesses information directly from ensembl using biomaRt (BioConductor).
```{r}
mouseIDTable <- Mouse2Human(MouseGenes = mouseIDTable$MGI)

# Convert symbol column ot dgelist
symbol = rownames(dgelist$counts)
dgelist$genes <- data.frame(Symbol = symbol)

# Add Hs homologue to dgelist
HsHmlg <- with(mouseIDTable, mouseIDTable$HGNC[match(row.names(dgelist), mouseIDTable$MGI)])
dgelist$genes$HsHmlg <- HsHmlg
dgelist$genes$HsHmlg[dgelist$genes$HsHmlg == ""] <- NA

# Convert row names to entrez gene ID
entrezid <- with(mouseIDTable, mouseIDTable$Entrez.Gene_ID[match(row.names(dgelist), mouseIDTable$MGI)])
entrezid[is.na(entrezid)] <- 'not_mapped'
rownames(dgelist$counts) <- entrezid
dgelist <- dgelist[!(row.names(dgelist) %in% 'not_mapped'), ] # get rid of unmapped entrez ids
```
#### Calculate library size 
Perform library correction with `DESeq2::rlog`
```{r}
dgelist$samples$lib.size <- colSums(dgelist$counts)
dgelist <- calcNormFactors(dgelist)
cts <- DESeq2::rlog(round(mstxi$counts), blind = TRUE)
```
#### Principal component analysis
prcomp PCA
```{r, fig.keep = TRUE}
pca.cts <- prcomp(t(cts))
scores=data.frame(pca.cts$x)
cols=ncol(scores)
scores=data.frame(scores,dgelist_groups)
colnames(scores)[cols+1]="cell"
colors <- c('darkviolet', 'cyan2', 'sienna1', 'seagreen3', 'brown4', 'violetred2', 'steelblue2', 'red1', 'darkgreen', 'black', 'royalblue')
points <- c(rep(19, 26))
```
__*Figure 5A*__
```{r}
scatterplot3d(scores[,1:3],
							color = colors[dgelist_groups],
							pch = points[dgelist_groups],
							lwd = 4,
							cex.symbols = 4,
							grid = T, box = F,
							type = 'h',
							xlab = 'PC1', ylab = 'PC2', zlab = 'PC3',
							main = 'RNA_3dPCA',
							angle = 215) 
legend("topright", legend = levels(dgelist_groups), pch = points, col = colors, ncol = 1, pt.cex = 0.5, cex = 0.5)

```
#### Differential gene expression
To determine if CMP subpopulations have expression profiles similar to progenitor or lineage-restricted populations, generate enrichment gene sets for each  
Create the design matrix for downstream (do not set CMP as intercept)
```{r}
designMat <- model.matrix(~ 0 + dgelist$samples$group)
colnames(designMat) <- levels(dgelist$samples$group)
designMat

# Apply GLM dispersion ------------------------------------------------------------

dgelist <- estimateGLMCommonDisp(dgelist, designMat)
dgelist <- estimateGLMTrendedDisp(dgelist, designMat, method = 'bin.spline')
dgelist <- estimateGLMTagwiseDisp(dgelist, designMat)
fit <- glmFit(dgelist, designMat)
```
Generate LSK, CMP, GMP, EB, MK gene sets
```{r}
bulk.contrasts <- makeContrasts(ERYvCMP = ERY - CMP, 
																ERYvGMP = ERY - GMP, 
																ERYvLSK = ERY - LSK, 
																ERYvMK = ERY - MK,
																
																GMPvCMP = GMP - CMP, 
																GMPvERY = GMP - ERY, 
																GMPvLSK = GMP - LSK, 
																GMPvMK = GMP - MK,
																
																LSKvCMP = LSK - CMP, 
																LSKvGMP = LSK - GMP, 
																LSKvERY = LSK - ERY, 
																LSKvMK = LSK - MK,
																
																MKvCMP = MK - CMP, 
																MKvGMP = MK - GMP, 
																MKvLSK = MK - LSK, 
																MKvERY = MK - ERY,
																
																CMPvERY = CMP - ERY, 
																CMPvGMP = CMP - GMP, 
																CMPvLSK = CMP - LSK, 
																CMPvMK = CMP - MK,
																
																levels = designMat)

bulk_population_expression_sets(bulk.contrasts, file.name = "LSK_CMP_GMP_EB_MK.gmt")
```
Now we have a gmt file in which each row contains a cell population (i.e., LSK, CMP, MEP, GMP, EB, MK) and a list of genes differentially expressed in that population versus all others

## Compare subpopulations to CMP in isolation
We want a list of genes upregulated in CMP subpopulations compared to CMP, regardless of their relative expression level in other populations. Therefore repeat DGE using only CMP subpopulation samples and set CMP as intercept
```{r, warning=FALSE}
file_list <- c(list.files(pattern = '_R', ignore.case = FALSE, recursive = FALSE, include.dirs = FALSE), list.files(pattern = '^CMP', recursive = FALSE, include.dirs = FALSE))
names(file_list) <- lapply(file_list, function(x) gsub('.genes.results', '', x))
mstxi <- tximport(files = file_list, type = 'rsem', txIn = FALSE, txOut = FALSE, geneIdCol = 3)
cts <- mstxi$counts

# Set reference sample 
dgelist_groups <- factor(c('C10', 'C10', 'C11', 'C11', 'C17', 'C17', 'C3s', 'C3s', 'CMP', 'CMP'))
dgelist_groups <- relevel(dgelist_groups, ref = "CMP")
dgelist <- DGEList(counts = cts, group = dgelist_groups)
keep <- rowSums(cpm(dgelist) > 5) >=2
dgelist <- dgelist[keep,]
symbol = rownames(dgelist$counts)
dgelist$genes <- data.frame(Symbol = symbol)

# Add Hs homologue to dgelist
HsHmlg <- with(mouseIDTable, mouseIDTable$HGNC[match(row.names(dgelist), mouseIDTable$MGI)])
dgelist$genes$HsHmlg <- HsHmlg
dgelist$genes$HsHmlg[dgelist$genes$HsHmlg == ""] <- NA
entrezid <- with(mouseIDTable, mouseIDTable$Entrez.Gene_ID[match(row.names(dgelist), mouseIDTable$MGI)])
entrezid[is.na(entrezid)] <- 'not_mapped'
rownames(dgelist$counts) <- entrezid
dgelist <- dgelist[!(row.names(dgelist) %in% 'not_mapped'), ] # get rid of unmapped entrez ids

```
Calculate library size
```{r}
dgelist$samples$lib.size <- colSums(dgelist$counts)
dgelist <- calcNormFactors(dgelist)
```
Estimate dispersion __with__ CMP defined as intercept
```{r, warning=FALSE}
designMat <- model.matrix(~dgelist$samples$group)
colnames(designMat) <- c('(Intercept)', levels(dgelist$samples$group)[2:length(levels(dgelist$samples$group))])

# Apply GLM dispersion ------------------------------------------------------------

dgelist <- estimateGLMCommonDisp(dgelist, designMat)
dgelist <- estimateGLMTrendedDisp(dgelist, designMat, method = 'bin.spline')
dgelist <- estimateGLMTagwiseDisp(dgelist, designMat)

# GLM differential expression testing ------------------------------------------------------------
fit <- glmFit(dgelist, designMat)


```
Create ranked gene files for CMP subpopulations

```{r}
lrt.c10 <-  glmLRT(fit, coef = 2)
gsea_ranking(lrt.c10, 'c10vCMP_2FC')

lrt.c11 <-  glmLRT(fit, coef = 3)
gsea_ranking(lrt.c11, 'c11vCMP_2FC')

lrt.c17 <-  glmLRT(fit, coef = 4)
gsea_ranking(lrt.c17, 'c17vCMP_2FC')

lrt.c3s <-  glmLRT(fit, coef = 5)
gsea_ranking(lrt.c3s, 'c3svCMP_2FC')
```
Perform [GSEA analysis](https://github.com/efheuston/CMPSubpopulation_dataAnalysis/blob/main/BulkRNA-SeqAnalysis.ipynb) to find (de)enrichment of genes differentially expressed in CMP subpopulations compared to LSK, CMP, GMP, EB, and MK.

